### PR TITLE
Refactoring without filewatching

### PR DESCRIPTION
**Pull request by tlienart**
Ok as discussed in #6 and #1, FileWatching actually makes  our life more difficult than it needs to by having a lot of async stuff everywhere (it turns out that these coroutines thing are actually quite hard to handle well apparently...)

Here's a suggestion for a (pretty big) refactoring of the code using simply a dictionary of watched files (`f_watcher`) which contains the time to last modification (like in JuDoc). The only `@async` is now the listener loop and that's it.

This works very well, is very fast and solves the issues #6, #1 and simplifies #5 . The code is also considerably simpler (I think).  But of course given the amount of change, I'd like you to check that you're happy with it before merging (*).

Additional goodies is that events are only triggered when the file is actually modified (and not for instance when you just select the file which is the case with FileWatching)

The changes do not not affect #2 (number of web sockets grows which we would still have to address separately. I have an idea of what we could do there which I'll outline there for suggestions (and I might need help).

(*) given that I've also modified the structure of the folder, depending on how happy you are to navigate GitHub's stuff you may want to pull this branch and look at it locally 🙂

### CODE REVIEW

**Comment by asprionj** [comment]
I'll have a closer look at it ASAP. I like simple solutions, and this really simplifies a lot. What I don't like is entangling stuff too much. We should adhere to the principle of "separation of concerns". The file watcher should be independent from the server, and only tell the server that this or that file has changed (e.g. via a callback as it is done now).
Of course, doing the watching in the main loop also catching Ctrl-C is a very convenient solution. But what if, in the future, there comes along a perfect package for file watching that we'd like to use? I'll  probably quickly try to separate the simple file-watching loop from the other stuff, I'll see soon enough if this leads anywhere...

**Comment by tlienart** [comment]
Ok so what I'm reading in your comment is primarily that the file-watcher should be separated.

Would it be an acceptable intermediate step if we wrapped the file-watching logic in a function? and then this answers your secondary question, we would just replace it by something else if something else comes up later on.

```julia
function watcher!(f_watcher, ws_tracker)
    while true
        # go over the files currently being watched
        for (f_path, time) ∈ f_watcher
            # check if the file still exists, if it doesn't remove from f_watcher
            isfile(f_path) || (delete!(f_watcher, f_path); continue)
            # retrieve the time of last modification
            cur_mtime = mtime(f_path)
            # check how it compares to the previously recorded time, if smaller --> update
            if f_watcher[f_path] < cur_mtime
                # update time of last modif
                f_watcher[f_path] = cur_mtime
                # trigger browser upgrade as appropriate
                if splitext(f_path)[2] == ".html"
                    ping_viewers!(ws_tracker[f_path])
                else
                    # e.g. css file may be needed by pages watched by any viewer so ping all
                    foreach(ping_viewers!, values(ws_tracker))
                end
            end
        end
        # yield (to the HTTP listener)
        sleep(0.1)
        end
    end
end
```

and then in the `serve` function:

```julia
try
    watcher!(f_watcher, ws_tracker)
catch err
    if isa(err, InterruptException)
        # stop_watching!(f_watcher)
        empty!(f_watcher)
        empty!(ws_tracker)
        println("\n✓ LiveServer shut down.")
    else
        throw(err)
    end
end
```

**Comment by tlienart** [comment]
Ok so  there were several problems that were nagging me. The websocket growth is now fixed thanks to your proposal to close after messaging (I've brought your code here and adapted it a bit). 

The final thing that was super annoying to fix is to forcibly close the listener; basically just closing the server _is not enough_. The problem is then that the async thread keeps living until you kill julia. But that means that if you were  to do 

```
serve()
CTRL+C
serve()
```

in the second case the initial listener was still used and  still being tied to the previous dictionaries (which are empty and stay that way). Therefore the updates didn't work.

Anyway long story short, HTTP.listen doesn't close properly  (I'll open an issue on HTTP.jl about this) and after trying multiple unsatisfactory  options, I resorted to the

```
if server.status != 4
    return 0
end
```

which does kill the async thread. This fixes the issue above (i.e. if you restart the server  later on, the update mechanism works).

Anyway so with the PR as it stands.

- the number of websockets doesn't grow anymore (as per your fix)
- killing the server actually kills all tasks
- restarting the server in the same julia session  "just works"

**Comment by asprionj** [comment]
I quickly draw an overview of what's in LiveServer and how the components relate to each other -- see [LiveServer_coroutines.pdf](https://github.com/asprionj/LiveServer.jl/files/3026370/LiveServer_coroutines.pdf). The communication between the three coroutines / tasks can be implemented in different ways. We'd probably have to go through

https://docs.julialang.org/en/v1/manual/parallel-computing/index.html

and then decide which way is best suited for our problem. I never did concurrent / parallel stuff in Julia, so I'll have to work myself into it. I like that stuff though, it's really interesting, so I'll happily do that ;-)

Of course, even when communicating via Channels or similar, we'd still have to pass (callback) functions to the two coroutines.

**Comment by tlienart** [comment]
Nice yeah I agree with this.

**Comment by asprionj** [comment]
I'd suggest to have a single task for the file watching, along the lines you proposed (checking every ~ 100ms for changes by checking `mtime`). And another one, namely `HTTP.listen` -- for which we've to find out how to properly kill it...

**Comment by tlienart** [comment]
So I already tried that but the issue is that 

* if it's a synchronous task, then it's identical to what is currently being done
* if it's asynchronous (and then you need to add a function `stop()` which would kill it), then there seem to have weird inconsistencies with print statements etc.

What I had before (which I think is what you're suggesting):

```julia
function serve(...; block=true/false)
  # the same stuff
  if block
      @sync watcher!(...)
   else
      @async watcher!(...)
   end
end

function stop(...)
  close(server)
  Base.throwto(watchertask, ...)
end
```

I can re-try this if you think it's worthwile but I'm not convinced about the usecase. If a user wants to be able to use Julia while the LiveServer is active, I think it would be better to completely isolate by doing the LiveServer thing in a separate processor and not use threads which will have to yield to whatever it is that the user may want to do.

Note that Python's simpleHTTPserver and browser-sync are both blocking?

**Comment by asprionj** [comment]
I have a working example for a simple file watcher (using `mtime`) with two async tasks and two corresponding channels for communication. The tasks are automatically and "recursively" stopped (i.e. their while-loops broken) when the channels are closed. Works fine; will integrate this in a branch to bring everything together.

**Comment by asprionj** [comment]
Ok, please have a look and test branch `simpleFW`. It's a whole new "philosophy" concerning file watching:

- The watcher is an argument to `serve()` (but defaults to `SimpleWatcher()`).
- The watcher has a `start()` method, which launches two tasks; one (T1) listening on a `Channel` (C1) to receive files to be watched, the other (T2) writing files that changed to another `Channel` (C2).
- A "listener task" (LT) is instantiated that listens to C2 and triggers the page reloads via websockets.
- C1 is passed to the file-serving callback to add all served files to the watcher.
- Tearing down all tasks is via the channels: closing C1 makes T1 return and is observed by T2, which then also returns. C2 is bound to T2 and thus is closed, which is detected by LT, which then returns.

I also added my version of a "safe shut-down". Closing the server and all websockets, then waiting until everything is actually in a `:closed` or `status == 6` state before printing the final "LiveServer shut down." message.

The entire thing works like a charm!

I only have one strange behaviour: when doing `serve()` and request a file with a browser, then `Ctrl-C` and `serve()` again. When visiting the page again with the browser, an `InvalidStateException("Channel is closed.", :closed)` is thrown (once for each file requested...?). The strange this is that I could verify that these errors are caused by instances of the `file_server!` request handler **from the previous run**! That is, the server, live update etc. works as it should, it's just annoying that there are errors when everything works fine. Probably these calls get stuck in the middle during the tear-down process? Some help tracking this down would be very welcome. Otherwise, I feel like this solution has none of the issues of earlier versions and has a quite flexible and extendable structure.

**Comment by tlienart** [comment]
Ok nice

I like:

* WatchedFile type
* Watcher as an argument

I'm not sure but potentially ok (potentially unnecessary complexity)

* Channels

A couple of issues:

* when clicking the "download link", hitting cancel then trying to modify the `index.html`, updates are not triggered anymore. If you reload the page manually, then modify updates are triggered again
* the server is still not properly closed: it seems to be a bit of a hit and miss (it did work for me the first time) but the second time, I see the 404 message. The 404 is good but it means that the listener is still active. What we should (always) see is the "Unable to connect" message. 

_____ 

I think we should take a decision in terms of the implementation though. I will port the ideas of the modular file watcher over, and try an idea to keep track of all websockets and close them, update this branch one last time and then maybe we decide which way we want this to go so that we don't let branches diverge too much? (I'm happy for you to take that last decision as it's your package and most of the thinking comes from you! very happy to try to help fix things around your design later on).

**Comment by asprionj** [comment]
Aren't the WS connections all correctly closed in branch `simpleFW`? I thought they are; I think once they're updated from a normal HTTP request, they're separate from the server/listener, and I do close them explicitly, cf. https://github.com/asprionj/LiveServer.jl/blob/202c5631976c901706091f3b7f0ddfb7ae4f9ad6/src/server.jl#L234-L238.

**Comment by asprionj** [comment]
Concerning `Channel`s: I thought this is the "official" way to communicate with tasks. Also, closing of a channel can be detected by tasks and they can return then. Or, the other way 'round, channels can be bound to tasks, so when a tasks closes, the channel also closes automatically. If you know / come up with a simpler solution, you're very welcome ;-)

**Comment by tlienart** [comment]
I think I'm going to close this PR for now as the discussion has drifted a bit, maybe merge your SimpleFW and we can move from there? I think to sumarise the points:

* I don't think we need to communicate between tasks
* HTTP.jl has a bug as per #10 where there are @async tasks that stay alive even after the server is shut down, this **must** be fixed otherwise we will always be at risk of weird errors if restarting the server etc
* We need tests

On the other hand

* The design for a standalone filewatcher is great and I think can be really convenient
* I think your idea of having a task for the filewatcher makes sense, but it can only work flawlessly if we can properly shut everything down (incl. HTTP)

I'll work on my fork separately as I think the design is simpler but it has flaws that I need to fix first, I just want to make it work and make sure I understand everything that's going on. Maybe we won't merge it but that's fine as long as I gain a better understanding of what goes on under the hood :-) 

Probably best to keep the discussion going on in separate issues as it may help us to track the solutions down better?

**Comment by asprionj** [comment]
Agreed, this discussion has almost no relation to the original PR anymore... ;-)
